src/TestResults/match-by-context.ts

```diff
@@ -0,0 +1,362 @@
+  const results = tContainer.childData.map((n) => toMatchResult(n.data[0], err));
+  tContainer.childContainers.forEach((c) => results.push(...toUnmatchedResults(c, err)));
+  return results;
+};
+
+type MessageType = 'context-mismatch' | 'match-failed' | 'unusual-match' | 'duplicate-test-name';
+const makeWarning = (fileName: string, verbose: boolean) => (
+  messageType: MessageType,
+  contextType: ContextType,
+  source: ContainerNode<ItBlock> | DataNode<ItBlock>,
+  assertion?: ContainerNode<TestAssertionStatus> | DataNode<TestAssertionStatus>,
+  extraReason?: string
+): void => {
+  if (!verbose) {
+    return;
+  }
+
+  const output = (message: string): void =>
+    console.warn(`[${fileName}] ${message} \n source=`, source, `\n assertion=`, assertion);
+
+  const blockType = contextType === 'container' ? 'describe' : 'test';
+  switch (messageType) {
+    case 'context-mismatch':
+      output(
+        `!! context mismatched !! ${contextType} nodes are different under "${source.name}": `
+      );
+      break;
+    case 'match-failed': {
+      output(`!! match failed !! ${blockType}: "${source.name}": ${extraReason} `);
+      break;
+    }
+    case 'duplicate-test-name': {
+      output(
+        `duplicate test names in the same describe block is not recommanded and might not be matched reliably: ${blockType} ${source.name}`
+      );
+      break;
+    }
+    case 'unusual-match': {
+      output(`unusual match: ${extraReason} : ${blockType} ${source.name}: `);
+      break;
+    }
+  }
+};
+
+const hasSameStructure = <N1 extends BaseNodeType, N2 extends BaseNodeType>(
+  list1: N1[],
+  list2: N2[]
+): boolean => list1.length === list2.length;
+
+/**
+ * match tests container with assertion container by their context structure.
+ * @param fileName
+ * @param tContainer
+ * @param aContainer
+ * @param verbose turns on/off the debugging warning messages.
+ */
+
+export const matchByContext = (
+  fileName: string,
+  tContainer: ContainerNode<ItBlock>,
+  aContainer: ContainerNode<TestAssertionStatus>,
+  verbose = true
+): TestResult[] => {
+  const warning = makeWarning(fileName, verbose);
+
+  /**
+   * this is where the actual match occurred. Each test and assertion container pair will try
+   * to match both its child-data and child-container list recursively.
+   *
+   * @param _tContainer
+   * @param _aContainer
+   * @returns a list of TestResult collected from all its children (data + containers)
+   */
+  const _matchByContext = (
+    _tContainer: ContainerNode<ItBlock>,
+    _aContainer: ContainerNode<TestAssertionStatus>
+  ): TestResult[] => {
+    // the match algorithm: first match by sequence if their have the same structure;
+    // then fallback to simple name-based matching. Upon each test block, it invokes the
+    // callback to process the matched results.
+    const matchList = <N1 extends NodeType<ItBlock>, N2 extends NodeType<TestAssertionStatus>>(
```

Comment:
Since it's a one-to-many relationship, I would still label the types and variables in a way that one can distinguish between code and assertions.

Reply:
I am not sure what you mean? you mean just the generic label?

Reply:
I was only talking about names like `N1/2` and `n`. Since the function obviously separates between `ItBlock` and `TestAssertionStatus` containers, I would also adapt the variable names to tell those two more easily apart.